### ISSUE TITLE
KCL1.8.10 - NPE in KinesisClientLibLeaseCoordinator.java:265

### ISSUE THREAD

**Issue by fstrudel** [awslabs/amazon-kinesis-client#476]
Hello,
Sorry for the lack of reproducible steps, we have not found when this happens, but we recently got a large number of NullPointerException from the library.
The machine running the software did not reboot or anything.

```
2018-12-13 16:38:30.083 ERROR 3664 --- [dProcessor-0002] c.a.s.k.c.lib.worker.InitializeTask      : Caught exception: 

java.lang.NullPointerException: null
	at com.amazonaws.services.kinesis.clientlibrary.lib.worker.KinesisClientLibLeaseCoordinator.getCheckpointObject(KinesisClientLibLeaseCoordinator.java:265) ~[amazon-kinesis-client-1.8.10.jar!/:na]
	at com.amazonaws.services.kinesis.clientlibrary.lib.worker.InitializeTask.call(InitializeTask.java:82) ~[amazon-kinesis-client-1.8.10.jar!/:na]
	at com.amazonaws.services.kinesis.clientlibrary.lib.worker.MetricsCollectingTaskDecorator.call(MetricsCollectingTaskDecorator.java:49) [amazon-kinesis-client-1.8.10.jar!/:na]
	at com.amazonaws.services.kinesis.clientlibrary.lib.worker.MetricsCollectingTaskDecorator.call(MetricsCollectingTaskDecorator.java:24) [amazon-kinesis-client-1.8.10.jar!/:na]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_191]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_191]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_191]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_191]
```

**Comment by sahilpalvia** [comment]
Are you trying to checkpoint in the `initialize` method in the IRecordProcessor implementation?

**Comment by anoopgarlapati** [comment]
@sahilpalvia No, we are not checkpointing in the `initialize` method of our IRecordProcessor implementation.

**Comment by sahilpalvia** [comment]
Do you have debug logs enabled? Can you see `No lease found with key <SHARD_ID>, returning null` in your logs?

Can you also verify if the shard the initialize task was called for had a lease in the DynamoDB table?

**Comment by fstrudel** [comment]
@sahilpalvia By searching the source code using https://github.com/awslabs/amazon-kinesis-client/search?q=No+lease+found+with+key&unscoped_q=No+lease+found+with+key I can see that this message is in `DynamoDBLeaseRefresher.java` which is only in 2.x versions, right?

Is https://github.com/awslabs/amazon-kinesis-client/blob/v1.x/src/main/java/com/amazonaws/services/kinesis/leases/impl/LeaseRenewer.java the corresponding 1.x version?
If yes, we can enable the logs for this class and see. Let us know.
Thanks!

**Comment by sahilpalvia** [comment]
The same message for 1.x exists in the [LeaseManager.java](https://github.com/awslabs/amazon-kinesis-client/blob/v1.x/src/main/java/com/amazonaws/services/kinesis/leases/impl/LeaseManager.java#L347). Turning on debug logs for the java package [com.amazonaws.services.kinesis.leases.impl](https://github.com/awslabs/amazon-kinesis-client/tree/v1.x/src/main/java/com/amazonaws/services/kinesis/leases/impl) would log all interactions that the KCL Worker has with the lease management table.

**Comment by adrian-baker** [comment]
We've just started seeing this logged in our production system, approximately 50,000 times in the last day.

Versions in use:

`com.amazonaws:amazon-kinesis-client:1.9.3
com.amazonaws:dynamodb-streams-kinesis-adapter:1.4.0
com.amazonaws:aws-java-sdk-core:1.11.492
com.amazonaws:aws-java-sdk-kinesis:1.11.438`

```
java.lang.NullPointerException: null
	at com.amazonaws.services.kinesis.clientlibrary.lib.worker.KinesisClientLibLeaseCoordinator.getCheckpointObject(KinesisClientLibLeaseCoordinator.java:265)
	at com.amazonaws.services.kinesis.clientlibrary.lib.worker.InitializeTask.call(InitializeTask.java:82)
	at com.amazonaws.services.kinesis.clientlibrary.lib.worker.MetricsCollectingTaskDecorator.call(MetricsCollectingTaskDecorator.java:49)
	at com.amazonaws.services.kinesis.clientlibrary.lib.worker.MetricsCollectingTaskDecorator.call(MetricsCollectingTaskDecorator.java:24)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by agarbutt** [comment]
@sahilpalvia having a similar issue, in my case i cannot find a lease in the table.

Architecture question, should each I have a lease table for each stream ARN?

**Comment by aggarwal** [comment]
There is clearly some kind of race condition here. It appears that the worker is trying to create a shard consumer for a lease that no longer exists. Submitted a code change to improve the logging to aid further debugging.

**Comment by anoopgarlapati** [comment]
Do we have any updates on this? This error keeps occurring occasionally and when it does that we have to kill that particular EC2 instance every time so that the worker releases its lease.

**Comment by aggarwal** [comment]
FYI this change was released in KCL 1.11.0 today.
